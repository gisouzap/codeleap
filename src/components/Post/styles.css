.post {
  border: 1px solid var(--color-grey-200);
  border-radius: 16px;
}

.post-header {
  background-color: var(--color-primary);
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  padding: 24px;
  border-radius: 16px 16px 0px 0px;
}

.post-header p {
  color: var(--color-white);
}

.post-icons {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  gap: 30px;
}

.post-icon {
  cursor: pointer;
  background: none;
  border: none;
  padding: none;
  transition: 200ms;
}

.post-icon:hover {
  opacity: 0.8;
}

.post-wrapper {
  padding: 24px;
  gap: 16px;
  display: flex;
  flex-direction: column;
}

.post-details p:nth-child(2) {
  color: var(--color-grey-300);
}

.post-details {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
}

.post-content {
  flex-direction: row;
  align-items: center;
}
