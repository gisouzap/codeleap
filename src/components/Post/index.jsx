import PropTypes from 'prop-types';

import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';

import Icon from '../Icon';
import Typography from '../Typography';

import './styles.css';

dayjs.extend(relativeTime);

function Post({ username, item, onDelete, onEdit }) {
  const time = dayjs(item.created_datetime).fromNow();

  return (
    <div className="post">
      <div className="post-header">
        <Typography variant="title">{item.title}</Typography>

        {username === item.username && (
          <div className="post-icons">
            <button type="button" onClick={onDelete} className="post-icon">
              <Icon name="trash" title="Delete" />
            </button>
            <button type="button" onClick={onEdit} className="post-icon">
              <Icon name="edit" title="Edit" />
            </button>
          </div>
        )}
      </div>
      <div className="post-wrapper">
        <div className="post-details">
          <Typography variant="bodyBold">{item.username}</Typography>
          <Typography variant="body">{time}</Typography>
        </div>

        <div className="post-content">
          <Typography variant="body">{item.content}</Typography>
        </div>
      </div>
    </div>
  );
}

export default Post;

Post.propTypes = {
  username: PropTypes.string,
  item: PropTypes.shape({
    title: PropTypes.string,
    username: PropTypes.string,
    content: PropTypes.string,
    id: PropTypes.number,
    created_datetime: PropTypes.string.isRequired,
  }),
  onDelete: PropTypes.func.isRequired,
  onEdit: PropTypes.func.isRequired,
};
