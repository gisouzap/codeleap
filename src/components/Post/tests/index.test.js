import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Post from '../index';

describe('Post', () => {
  const username = 'codeleapUser';
  const item = {
    title: 'CodeLeap',
    content: 'Hello world',
    username: 'codeleapUser',
    created_datetime: '"2023-02-12T15:42:38.421Z"',
  };

  const handleDelete = jest.fn();
  const handleEdit = jest.fn();

  it('should render and match the snapshot', () => {
    const { container } = render(
      <Post
        username={username}
        item={item}
        onDelete={handleDelete}
        onEdit={handleEdit}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('should call handleEdit when delete button is clicked', () => {
    const { getByTitle } = render(
      <Post
        username={username}
        item={item}
        onDelete={handleDelete}
        onEdit={handleEdit}
      />
    );

    const editBtn = getByTitle('Edit');

    fireEvent.click(editBtn);

    expect(handleEdit).toHaveBeenCalledTimes(1);
  });

  it('should call handleDelete when delete button is clicked', () => {
    const { getByTitle } = render(
      <Post
        username={username}
        item={item}
        onDelete={handleDelete}
        onEdit={handleEdit}
      />
    );

    const deleteBtn = getByTitle('Delete');

    fireEvent.click(deleteBtn);

    expect(handleDelete).toHaveBeenCalledTimes(1);
  });
});
