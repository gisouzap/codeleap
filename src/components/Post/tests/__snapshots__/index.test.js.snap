// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post should render and match the snapshot 1`] = `
<div>
  <div
    class="post"
  >
    <div
      class="post-header"
    >
      <p
        class="title"
      >
        CodeLeap
      </p>
      <div
        class="post-icons"
      >
        <button
          class="post-icon"
          type="button"
        >
          <img
            src="test-file-stub"
            title="Delete"
          />
        </button>
        <button
          class="post-icon"
          type="button"
        >
          <img
            src="test-file-stub"
            title="Edit"
          />
        </button>
      </div>
    </div>
    <div
      class="post-wrapper"
    >
      <div
        class="post-details"
      >
        <p
          class="bodyBold"
        >
          codeleapUser
        </p>
        <p
          class="body"
        >
          NaN years ago
        </p>
      </div>
      <div
        class="post-content"
      >
        <p
          class="body"
        >
          Hello world
        </p>
      </div>
    </div>
  </div>
</div>
`;
